connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala

```diff
@@ -508,21 +451,135 @@ object KustoWriter {
+          batchIdForTracing,
+          currentBlobIndex.incrementAndGet())
+
+      }
+    }
+    if (parameters.writeOptions.ensureNoDupBlobs && taskMap.size() > 0) {
+      taskMap.forEach((blobUUID, blobWriter) => {
+        ingest(
+          blobWriter,
+          parameters,
+          ingestionProperties,
+          flushImmediately = false,
+          kustoClient,
+          partitionsResults,
+          batchIdForTracing,
+          currentBlobIndex.incrementAndGet())
+      })
+    }
+  }
+
+  def ingest(
+      blobResource: BlobWriteResource,
+      parameters: KustoWriteResource,
+      ingestionProperties: IngestionProperties,
+      flushImmediately: Boolean = false,
+      kustoClient: ExtendedKustoClient,
+      partitionsResults: CollectionAccumulator[PartitionResult],
+      batchIdForTracing: String,
+      blobIndexInBatch: Int): Unit = {
+    val size = blobResource.csvWriter.getCounter
+    val sas = blobResource.sas
+    val partitionId = TaskContext.getPartitionId
+    var props = ingestionProperties
+    val blobUri = blobResource.blob.getStorageUri.getPrimaryUri.toString
+    if (parameters.writeOptions.ensureNoDupBlobs || (!props.getFlushImmediately && flushImmediately)) {
+      // Need to copy the ingestionProperties so that only this blob ingestion will be effected
+      props = SparkIngestionProperties.cloneIngestionProperties(ingestionProperties)
+    }
+    if (parameters.writeOptions.ensureNoDupBlobs) {
+//      if (blobIndexInBatch == 2 && TaskContext.get().taskAttemptId() % 2 == 0 && exceptionCount
+//          .get() < 2) {
+//        exceptionCount.incrementAndGet()
+//        KDSU.logWarn(
+//          className,
+//          "********************************************Forcing an exception to test the retry mechanism********************************************")
+//        throw new Exception("***Test Exception****")
+//      }
+      // The Key change is here
+      val tag = KDSU.getDedupTagsPrefix(
+        parameters.writeOptions.requestId,
+        s"${blobIndexInBatch.toString}_${partitionId}_${StringUtils.defaultIfBlank(batchIdForTracing, "0")}")
+      KDSU.logInfo(
+        className,
+        s"With ensureNoDupBlobs in partition: ${TaskContext.getPartitionId}, " +
+          s"for requestId: '${parameters.writeOptions.requestId}, tag: $tag , blobIndexInBatch: $blobIndexInBatch " +
+          s"and batchIdForTracing: ${StringUtils.defaultIfBlank(batchIdForTracing, "0")}")
+      val ingestIfNotExist = new util.ArrayList[String]
+      ingestIfNotExist.addAll(props.getIngestIfNotExists)
+      val ingestBy: util.List[String] = new util.ArrayList[String]
+      ingestBy.addAll(props.getIngestByTags)
+      ingestBy.add(tag)
+      ingestIfNotExist.add(tag)
+      props.setIngestByTags(ingestBy)
+      props.setIngestIfNotExists(ingestIfNotExist)
     }
+    if (!props.getFlushImmediately && flushImmediately) {
+      props.setFlushImmediately(true)
+    }
+    val ingestClient = KustoClientCache
```

Comment:
why do we need to get the client from cache everytime instead of passing it in params?